src/NuGet.Core/NuGet.CommandLine.XPlat/Program.cs

```diff
@@ -127,25 +130,83 @@ public static int Main(string[] args)
                         }
                     }
 
-                    var restoreExitCode = 0;
+                    // Run restores
+                    var isParallel = !disableParallel.HasValue();
+                    var maxTasks = isParallel ? 16 : 1;
+
+                    if (isParallel)
+                    {
+                        Log.LogVerbose(string.Format(
+                            CultureInfo.CurrentCulture,
+                            Strings.Log_RunningParallelRestore,
+                            maxTasks));
+                    }
+                    else
+                    {
+                        Log.LogVerbose(Strings.Log_RunningNonParallelRestore);
+                    }
+
+                    var localCaches = new Dictionary<string, NuGetv3LocalRepository>(StringComparer.Ordinal);
+
+                    var success = true;
+                    var restoreTasks = new List<Task<int>>(maxTasks);
 
                     foreach (var inputPath in inputValues)
                     {
-                        var currentExitCode = await ExecuteRestore(
+                        // Global folder
+                        // Load settings based on the current project path.
+                        var projectDir = Path.GetDirectoryName(inputPath);
+                        var settings = Settings.LoadDefaultSettings(projectDir,
+                            configFileName: null,
+                            machineWideSettings: null);
+
+                        var globalFolderPath = string.Empty;
+                        if (packagesDirectory.HasValue())
+                        {
+                            globalFolderPath = packagesDirectory.Value();
+                        }
+                        else
+                        {
+                            globalFolderPath = SettingsUtility.GetGlobalPackagesFolder(settings);
+                        }
+
+                        // Find the shared local cache for globalFolderPath
+                        // The global folder may differ between projects
+                        NuGetv3LocalRepository localCache;
+                        if (!localCaches.TryGetValue(globalFolderPath, out localCache))
+                        {
+                            localCache = new NuGetv3LocalRepository(globalFolderPath);
+                            localCaches.Add(globalFolderPath, localCache);
+                        }
+
+                        // Throttle and wait for a task to finish if we have hit the limit
+                        if (restoreTasks.Count == maxTasks)
+                        {
+                            success &= await CompleteTask(restoreTasks);
+                        }
+
+                        // Start a new restore
+                        var task = ExecuteRestore(
```

Comment:
Couldn't `ExecuteRestore` just return the Success bool on the `RestoreResult`? Right now we are going from bool to exit code (`int`) to bool to exit code.